google/cloud/storage/internal/rest_request_builder_test.cc

```diff
@@ -0,0 +1,204 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/internal/rest_request_builder.h"
+#include "google/cloud/storage/internal/generic_request.h"
+#include "google/cloud/storage/well_known_headers.h"
+#include "google/cloud/storage/well_known_parameters.h"
+#include <gmock/gmock.h>
+
+namespace google {
+namespace cloud {
+namespace storage {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+namespace {
+
+using google::cloud::rest_internal::RestRequest;
+using ::testing::Eq;
+using ::testing::UnorderedElementsAre;
+
+TEST(RestRequestBuilderTest, WellKnownParameterRequest) {
+  class WellKnownParameterRequest
+      : public GenericRequest<WellKnownParameterRequest, Deleted, MaxResults,
+                              Prefix, Projection, UserProject> {};
+
+  WellKnownParameterRequest request;
+  request.set_option(Deleted(true));
+  request.set_option(MaxResults(42));
+  request.set_option(Prefix("my_prefix"));
+  request.set_option(Projection("my_projection"));
+  request.set_option(UserProject("my_project"));
+
+  RestRequestBuilder builder("service/path");
+  request.AddOptionsToHttpRequest(builder);
```

Comment:
I am not sure this helps: `GenericRequest<>` now has a `.ForEachOption()` member function that calls `operator()` on each of its `Option` values.  This is actually what gets used in `AddOptionsToHttpRequest`:

https://github.com/googleapis/google-cloud-cpp/blob/d405ec7691611a8a687d2bf9f69dcc21a9171455/google/cloud/storage/internal/generic_request.h#L154-L158

Maybe you want to skip the extra level of indirection.